### PR TITLE
Updated transform Gizmos

### PR DESCRIPTION
**Pull request by W2Wizard**
# Some good looking Gizmos!
Since #49 was closed for no given reason I took my time to recreate it based on @DaveTheGameDev's work.

This PR wasn't aiming at fixing some bugs regarding the gizmo but during the work I found the cause for #323 so atleast theres some upside to this all :D

## Files | Includes all assets *.flax
[Gizmo.zip](https://github.com/FlaxEngine/FlaxEngine/files/6176222/Gizmo.zip)

### Known issues:
Center translation currently does not work at all.

**Example**:

https://user-images.githubusercontent.com/63303990/111080755-33c56200-8500-11eb-9275-398f4d32745e.mp4

**Translation**:
image

**Rotation**:
image

**Scale**:
image

### CODE REVIEW

**Comment by W2Wizard** [comment]
Gizmo can now fully adjust to the view position 🥳 

https://user-images.githubusercontent.com/63303990/111177386-7479b580-85aa-11eb-937f-b11658b372c0.mp4

Also those merge conflicts are just empty spaces lol

**Comment by W2Wizard** [comment]
Discussed the uses of the view alignment and it was decided that it would cause too much confusion so its gone now, but the main goal was to actually improve the gizmos and that has been achieved

**Comment by mafiesto4** [comment]
I have some problems merging this one. Specifically modified assets using LFS are failing to fetch:
```
$ git lfs fetch --all
fetch: 988 object(s) found, done.
fetch: Fetching all references...
[8d8f319dcd39842cb991a4e8e67e4823c7159a493a1aec757232077e5d56ad9f] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
[8743967dc27d5839b88dfa1e599cbc9b290427afd6b65b2424ffa5218b330beb] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
[a6096ef866920a861ec91f0660999a9ea07b2fbd26bcf52ba2c0112ba5e02336] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
[624ab02e6eda3adb6b96a7a8c3d666cb78da7b0b47b752fcd232a4280bc60078] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
[63b70397980f49d234cfcdcf5decdafe4d8032b513a15cfa32bd2758ffa0c088] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
[96910bc6b88f6f25f2ec96fcecd7bd7e763074e54fa0fa4789977fd31e81efdc] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
[80f6f7303407070b54181c7ab008db755088f4eba899475abcac0ff92d5b5422] Object does not exist on the server or you don't have permissions to access it: [404] Object does not exist on the server or you don't have permissions to access it
error: failed to fetch some objects from 'https://gitlab.flaxengine.com/flax/flaxengine.git/info/lfs'
```

Could you upload those assets here or share them so I could add them manually?

**Comment by W2Wizard** [comment]
I already added them ontop , Gizmo.zip.

**Comment by W2Wizard** [comment]
Regarding the merge conflitcs, my changes are supposed the overwrite the mentioned conflicts (for TransformGizmoBase mainly) the other conflicts are just empty spaces :P

File: Source/Engine/Core/Math/BoundingBox.cs [hunk 1]
```diff
@@ -414,6 +414,31 @@ public static BoundingBox Merge(BoundingBox value1, BoundingBox value2)
...
+            Vector3.Max(ref Maximum, ref point, out result.Maximum);
+            return result;
+        }
+
+        /// <summary>
+        /// Creates the bounding box that is offseted by the given vector. Adds the offset value to minimum and maximum points.
+        /// </summary>
+        /// <param name="offset">The offset.</param>
+        /// <returns>The result.</returns>
+        public BoundingBox MakeOffsetted(Vector3 offset)
```
Comment (stefnotch) [inline Source/Engine/Core/Math/BoundingBox.cs:434]: Those are taken from the C++ side 👍  https://github.com/FlaxEngine/FlaxEngine/blob/7c9e170a40f748bc2738d6ab40649fbf13c742f9/Source/Engine/Core/Math/BoundingBox.cpp#L58

File: Source/Editor/Viewport/EditorViewport.cs [hunk 2]
```diff
@@ -143,6 +144,11 @@ public void Clear()
         private Vector2 _mouseDeltaLast;
         private Vector2[] _deltaFilteringBuffer = new Vector2[FpsCameraFilteringFrames];
 
+        /// <summary>
+        /// Mouse delta (x = Left, Y = Right)
+        /// </summary>
+        protected Vector2 _mouseDelta;
```
Comment (stefnotch) [inline Source/Editor/Viewport/EditorViewport.cs:150]: Where did that come from and is it still used?

File: Source/Editor/Gizmo/TransformGizmoBase.Settings.cs [hunk 3]
```diff
@@ -7,56 +7,51 @@ namespace FlaxEditor.Gizmo
...
+        /// The inner minimum of the multiscale
+        /// </summary>
+        private const float InnerExtend = AxisOffset + 0.5f;
+
+        /// <summary>
+        /// The outer maximum of the multiscale
+        /// </summary>
+        private const float OuterExtend = AxisOffset * 3.5f;
+
+        private BoundingBox XAxisBox = new BoundingBox(new Vector3(-AxisThickness), new Vector3(AxisThickness)).MakeOffsetted(AxisOffset * Vector3.UnitX).Merge(AxisLength * Vector3.UnitX);
```
Comment (stefnotch) [inline Source/Editor/Gizmo/TransformGizmoBase.Settings.cs:45]: For those wondering what this does, it creates a cube with the size `AxisThickness`, then moves it along the axis (`AxisThickness`) and finally makes it really long (`AxisLength`)

**Review by stefnotch — APPROVED** [review]
Looks good to me!

File: Source/Editor/Viewport/EditorViewport.cs (hunk 2 above)
Comment (W2Wizard) [reply Source/Editor/Viewport/EditorViewport.cs:150]: This I think was some left over stuff from an attempt to fix some dragging issue, its actually duplicated now.

Comment (W2Wizard) [reply Source/Editor/Viewport/EditorViewport.cs]: Fixed.

**Comment by W2Wizard** [comment]
We're about to celebrate this PR's 1 year anniversary in a month

**Comment by W2Wizard** [comment]
@mafiesto4 Would you have some time to review this, its been so long I don't remember now what exactly is important and so on to fix the conflict anymore.

giphy

**Comment by RuanLucasGD** [comment]
I'm looking forward to using the new gizmos, they look so much better. Any precision on when it will be added? will it come in version 1.5? @W2Wizard @mafiesto4

**Comment by W2Wizard** [comment]
pls merge

**Comment by RuanLucasGD** [comment]
@mafiesto4 @Tryibion

**Comment by W2Wizard** [comment]
Happy new year! Can we merge ?

**Comment by DaveTheGameDev** [comment]
I closed it for this very reason. No response from the devs.

**Comment by W2Wizard** [comment]
I can understand if something is a low priority feature compared to other things but leaving a PR hanging, that has clearly some interest from others, for like 2 years is getting a bit silly

**Comment by mafiesto4** [comment]
image

**Comment by mafiesto4** [comment]
Merged 🎉  f8aa1cd5f8fdeea9129a7a97c807f75eccf3df4f
(center sphere for scale/translate is disabled for now as it needs some fixing)

**Pull request closed without merging**